build.gradle.kts

```diff
@@ -14,20 +16,19 @@ allprojects {
     group = PUBLISHING_GROUP
 }
 
-val detektFormatting = libs.detekt.formatting
+gitHooks {
+    setHooks(mapOf("pre-commit" to "check"))
+}
 
 subprojects {
     apply {
         plugin("io.gitlab.arturbosch.detekt")
+        plugin("org.jlleitschuh.gradle.ktlint")
```

Comment:
I'd rather prefer to stick with detekt formatting. What was wrong with it?

Reply:
I made a change in one of the kotlin source files that is not consistent with the ktlint enforced code style and `./gradlew check` didn't detect this issue. Integrating directly with ktlint fixed this.

How do you confirm that ktlint is being applied to the kotlin source code through the detekt formatting plugin?

Reply:
You're right that the formatting is not working fine. Please add this section at the bottom of `detekt.yml` and you will be formatting with the `check` step:

```
formatting:
  active: true
  android: false
  autoCorrect: true
  AnnotationOnSeparateLine:
    active: true
    autoCorrect: true
  AnnotationSpacing:
    active: true
    autoCorrect: true
  ArgumentListWrapping:
    active: true
    autoCorrect: true
    indentSize: 4
    maxLineLength: 120
  BlockCommentInitialStarAlignment:
    active: false
    autoCorrect: true
  ChainWrapping:
    active: true
    autoCorrect: true
  CommentSpacing:
    active: true
    autoCorrect: true
  CommentWrapping:
    active: false
    autoCorrect: true
    indentSize: 4
  DiscouragedCommentLocation:
    active: false
    autoCorrect: true
  EnumEntryNameCase:
    active: true
    autoCorrect: true
  Filename:
    active: true
  FinalNewline:
    active: true
    autoCorrect: true
    insertFinalNewLine: true
  FunKeywordSpacing:
    active: false
    autoCorrect: true
  FunctionReturnTypeSpacing:
    active: false
    autoCorrect: true
  FunctionSignature:
    active: false
    autoCorrect: true
    forceMultilineWhenParameterCountGreaterOrEqualThan: 2147483647
    functionBodyExpressionWrapping: 'default'
    maxLineLength: 120
    indentSize: 4
  FunctionStartOfBodySpacing:
    active: false
    autoCorrect: true
  FunctionTypeReferenceSpacing:
    active: false
    autoCorrect: true
  ImportOrdering:
    active: true
    autoCorrect: true
    layout: '*,java.**,javax.**,kotlin.**,^'
  Indentation:
    active: true
    autoCorrect: true
    indentSize: 4
  KdocWrapping:
    active: false
    autoCorrect: true
    indentSize: 4
  MaximumLineLength:
    active: true
    maxLineLength: 120
    ignoreBackTickedIdentifier: false
  ModifierListSpacing:
    active: false
    autoCorrect: true
  ModifierOrdering:
    active: true
    autoCorrect: true
  MultiLineIfElse:
    active: true
    autoCorrect: true
  NoBlankLineBeforeRbrace:
    active: true
    autoCorrect: true
  NoBlankLinesInChainedMethodCalls:
    active: true
    autoCorrect: true
  NoConsecutiveBlankLines:
    active: true
    autoCorrect: true
  NoEmptyClassBody:
    active: true
    autoCorrect: true
  NoEmptyFirstLineInMethodBlock:
    active: true
    autoCorrect: true
  NoLineBreakAfterElse:
    active: true
    autoCorrect: true
  NoLineBreakBeforeAssignment:
    active: true
    autoCorrect: true
  NoMultipleSpaces:
    active: true
    autoCorrect: true
  NoSemicolons:
    active: true
    autoCorrect: true
  NoTrailingSpaces:
    active: true
    autoCorrect: true
  NoUnitReturn:
    active: true
    autoCorrect: true
  NoUnusedImports:
    active: true
    autoCorrect: true
  NoWildcardImports:
    active: true
    packagesToUseImportOnDemandProperty: 'java.util.*,kotlinx.android.synthetic.**'
  NullableTypeSpacing:
    active: false
    autoCorrect: true
  PackageName:
    active: true
    autoCorrect: true
  ParameterListSpacing:
    active: false
    autoCorrect: true
  ParameterListWrapping:
    active: true
    autoCorrect: true
    maxLineLength: 120
  SpacingAroundAngleBrackets:
    active: true
    autoCorrect: true
  SpacingAroundColon:
    active: true
    autoCorrect: true
  SpacingAroundComma:
    active: true
    autoCorrect: true
  SpacingAroundCurly:
    active: true
    autoCorrect: true
  SpacingAroundDot:
    active: true
    autoCorrect: true
  SpacingAroundDoubleColon:
    active: true
    autoCorrect: true
  SpacingAroundKeyword:
    active: true
    autoCorrect: true
  SpacingAroundOperators:
    active: true
    autoCorrect: true
  SpacingAroundParens:
    active: true
    autoCorrect: true
  SpacingAroundRangeOperator:
    active: true
    autoCorrect: true
  SpacingAroundUnaryOperator:
    active: true
    autoCorrect: true
  SpacingBetweenDeclarationsWithAnnotations:
    active: true
    autoCorrect: true
  SpacingBetweenDeclarationsWithComments:
    active: true
    autoCorrect: true
  SpacingBetweenFunctionNameAndOpeningParenthesis:
    active: false
    autoCorrect: true
  StringTemplate:
    active: true
    autoCorrect: true
  TrailingCommaOnCallSite:
    active: false
    autoCorrect: true
    useTrailingCommaOnCallSite: true
  TrailingCommaOnDeclarationSite:
    active: false
    autoCorrect: true
    useTrailingCommaOnDeclarationSite: true
  TypeArgumentListSpacing:
    active: false
    autoCorrect: true
  TypeParameterListSpacing:
    active: false
    autoCorrect: true
  UnnecessaryParenthesesBeforeTrailingLambda:
    active: false
    autoCorrect: true
  Wrapping:
    active: true
    autoCorrect: true
    indentSize: 4
```

Reply:
> How was this config derived?

That's from the official Detekt repo.
https://github.com/detekt/detekt/blob/main/detekt-formatting/src/main/resources/config/config.yml

or you can invoke `./gradlew detektGenerateConfig`

This whole point was already discussed here:
- https://github.com/cortinico/kotlin-android-template/pull/35

I'd rather keep detekt-formatting here as I'm one of the maintainer of Detekt, and you can do static analysis and formatting with just one gradle plugin rather than two. We've also surveyed the community and seems like a significant group of users are using this wrapper: 
- https://github.com/detekt/detekt/discussions/5997
So I'd rather keep it here for now.

Reply:
@cortinico - *bump* - I'd love to get an answer to this one last question so I can get this PR completed and merged.

Reply:
Sorry for the late reply but I was (and I still am) quite busy at the moment so I'll be slower to respond. This PR is still in my inbox so I intend to to merge it 👍 

> Is there a gradle detekt task which will autofix formatting issues?

If `autoCorrect` is set to `true` in the Detekt config's file, detekt should run ktlint's autoCorrection on those issues

Reply:
> I'd love to get an answer to this one last question so I can get this PR completed and merged.

Following up here, if you run `./gradlew detekt --auto-correct` it will format those files

Reply:
Thanks for the response!

I've added some extra spaces on line 8 of this file:

```
package com.ncorti.kotlin.template.library

object FactorialCalculator {
    private const val MAX_FACTORIAL_64BIT = 20

    tailrec fun computeFactorial(input: Long, temp: Long = 1L): Long =
        when {
            input < 0 -> error(        "Factorial is not defined for negative numbers")
            input > MAX_FACTORIAL_64BIT -> error("Only a factorial up to 20 can fit in a 64-bit Long")
            input == 0L -> temp
            else -> computeFactorial(input - 1, temp * input)
        }
}
```

If I run `./gradlew detekt` I get the following errors:
```
/Users/mchr3k/Workspaces/kotlin-android-template/library-kotlin/src/main/java/com/ncorti/kotlin/template/library/FactorialCalculator.kt:8:32: Unnecessary long whitespace [NoMultipleSpaces]
/Users/mchr3k/Workspaces/kotlin-android-template/library-kotlin/src/main/java/com/ncorti/kotlin/template/library/FactorialCalculator.kt:8:31: Unexpected spacing after "(" [SpacingAroundParens]
```

This is despite `detekt.yml` specifying `autoCorrect: true` for both of these rules:
```
...
  NoMultipleSpaces:
    active: true
    autoCorrect: true
...
  SpacingAroundParens:
    active: true
    autoCorrect: true
...
```

If I run `./gradlew detekt --auto-correct` then the detekt task still fails:
```
> Task :library-kotlin:detekt FAILED
Property 'naming>BooleanPropertyNaming>ignoreOverridden' is deprecated. This configuration is ignored and will be removed in the future.
Property 'style>MandatoryBracesIfStatements' is deprecated. Use `BracesOnIfStatements` with `always` configuration instead.
Property 'style>ForbiddenComment>customMessage' is deprecated. Use `comments` and provide `reason` against each `value`..
Property 'style>ForbiddenComment>values' is deprecated. Use `comments` instead, make sure you escape your text for Regular Expressions..
Property 'naming>FunctionNaming>ignoreOverridden' is deprecated. This configuration is ignored and will be removed in the future.
Property 'naming>FunctionParameterNaming>ignoreOverridden' is deprecated. This configuration is ignored and will be removed in the future.
Property 'naming>ConstructorParameterNaming>ignoreOverridden' is deprecated. This configuration is ignored and will be removed in the future.
Property 'naming>VariableNaming>ignoreOverridden' is deprecated. This configuration is ignored and will be removed in the future.
/Users/mchr3k/Workspaces/kotlin-android-template/library-kotlin/src/main/java/com/ncorti/kotlin/template/library/FactorialCalculator.kt:8:31: Unexpected spacing after "(" [SpacingAroundParens]
```

But the formatting issue has been fixed, and so `./gradlew detekt` now succeeds.

Questions:
* Why is the `--auto-correct` command line switch needed when `detekt.yml` specifies `autoCorrect: true`?
* Why does `Task :library-kotlin:detekt` fail when the only detected issue has been autocorrected? It looks to me like this is a bug caused by the same formatting issue triggering multiple rules and so the initial task failure outputs an error for `NoMultipleSpaces` and `SpacingAroundParens`, but the `--auto-correct` run only outputs an error related to `SpacingAroundParens`.

Reply:
The `--auto-correct` invocation will fail if the formatting is not valid. If you run it again, it should be true. The formatter needs a way to report if the formatting is valid or not, and that's what Detekt is using (build status).

> Why is the --auto-correct command line switch needed when detekt.yml specifies autoCorrect: true?

Because `autoCorrect: true` has granularity at the rule level. So you could enable/disable it for each rule or ruleset. But it won't actually trigger the autocorrection unless   the user specifies the `--auto-correct` flag.

Reply:
Great, that makes sense, thanks! In that case I think this point is now resolved.